### ISSUE TITLE
Garbage collector behavior on invalid ownerReferences for existing uids across namespaces and across kinds is non-deterministic

### ISSUE THREAD

**Issue by caesarxuchao** [kubernetes/kubernetes#65200]
If an object with a given uid is already in the garbage collector uid map, a child object created with an ownerReference pointing to that uid is not treated as having a non-existent parent, even if:
1. the ownerReference refers to a cluster-scoped kind and no cluster-scoped kind with that name exists
2. the child is namespaced, and the ownerReference refers to a namespaced kind and no namespace-scoped kind with that name exists in the same namespace
3. the child is cluster-scoped, and the ownerReference refers to a namespaced kind (since the namespace of the parent is ambiguous, it should be treated as not existing)

Original description follows
=============

Forked from https://github.com/kubernetes/kubernetes/pull/63386#discussion_r193257021.

Garbage collector should work for three cases: (a) cluster-scoped owner with namespaced dependents, (b) namespaced owner and namespaced dependents that are in the same namespace, and (c) cluster-scoped owner with cluster-scoped dependents.

Garbage collector should NOT work for the other two cases: (c) namespaced owner with cluster-scoped dependents. (d) owner and dependents that are in different namespaces. Today, GC sometimes work in these two cases. It's a bug for two reasons:

* This weakens our security guarantees. Users might exploit this bug to grant other users delete permissions.
* Users might rely on the unsupported behavior.

We can add extra checks in the GC controller to make it never work for case (c) and (d).

cc @lavalamp @liggitt @deads2k

**Comment by tpepper** [comment]
The discussion in #63386 is unclear to me.  While that is 1.11 milestone material, is the content of this issue here a lower priority and deferred to the next milestone and perhaps later cherry-picking?  Given the delay, I'd like to assume it is not critical-urgent, but rather than assume...

@kubernetes/sig-api-machinery-bugs can folks comment?

**Comment by liggitt** [comment]
This is longstanding and not a 1.11 blocker.

**Comment by roytman** [comment]
I work with K8s 1.13 and observed that GC works when 

> (d) owner and dependents that are in different namespaces

If required, I can provide a simple test example.

**Comment by liggitt** [comment]
> I work with K8s 1.13 and observed that GC works when
> 
> > (d) owner and dependents that are in different namespaces

It does not work reliably. If you set up a relationship like that, then restart the controller manager, the dependent will usually be deleted (because the uid-indexed owner cache is not populated yet)

**Comment by roytman** [comment]
I mean that in my case the dependent is deleted without restart the controller manager. It is quite reproducible.

**Comment by liggitt** [comment]
> I mean that in my case the dependent is deleted without restart the controller manager. It is quite reproducible.

That is the expected/desired behavior, but is racy, so *sometimes* the cross-namespace relationship does not result in deletion until restart. Because it is racy, environmental factors like load and performance (both of the API server and controller manager) can affect results.

**Comment by roytman** [comment]
I see, thanks

**Comment by liggitt** [comment]
xref #72192

**Comment by jennybuckley** [comment]
I think this bug is a little more dangerous than the two reasons given in the original description, since due to virtual nodes, having one illegal cross namespace dependent of X can cause  the GC to delete other, legal dependents of X, occasionally

**Comment by tedyu** [comment]
Should we pay attention to the performance impact due to the extra checks in the GC controller ?
Looking under pkg/controller/garbagecollector , I don't see any benchmark test.

**Comment by tedyu** [comment]
In GarbageCollector#attemptToDeleteItem :
```
		for _, dep := range deps {
			if dep.isDeletingDependents() {
...
				if _, err := gc.patch(item, patch, gc.unblockOwnerReferencesJSONMergePatch); err != nil {
					return err
				}
				break
```
I think we should continue with other dependents performing the unblocking (instead of breaking out of the loop).

**Comment by josiahbjorgaard** [comment]
Bug triage for release 1.17 here. Since this issue has been tagged for the milestone, we would like to check it's status. Code freeze is on Nov. 19. Will this issue be resolved before then? If not, we should move it out of the milestone.

**Comment by josiahbjorgaard** [comment]
@liggitt Could you confirm the status of this issue with regard to v1.17 and the recent PR referencing it? Should we have that PR in the milestone?

**Comment by liggitt** [comment]
no, that PR is not planned to merge as is. the changes required to fix this issue are too risky for this point in the release, will defer to 1.18

/milestone v1.18

**Comment by gianarb** [comment]
Hello @liggitt !
Bug Triage team here for the 1.18 release. This is a friendly reminder that code freeze is scheduled for 5 March. Is this issue still intended for milestone 1.18, or should it extend to the next one? Thanks in advance!

**Comment by bbrowning** [comment]
My team was just bitten by an otherwise-valid ownerRef from a namespace-scoped object to an object in a different namespace. This caused some things to get deleted that definitely should not have been deleted leading to all kinds of headaches. It's bad on us for creating this invalid ownerReference in one of our controllers, but it's worse on Kubernetes for letting us set it and then nuking the owner during GC just because the owned thing has a cross-namespace ownerRef.

This lets any user elevate their view permissions on an resource to delete permissions just by creating any other resource with careful ownerReferences and waiting on GC.

**Comment by liggitt** [comment]
> nuking the owner during GC

To be clear, an ownerRef does not cause deletion of the owner

**Comment by liggitt** [comment]
I agree this is important to fix. I am looking into potential mitigations to make behavior in the presence of these ownerRefs more deterministic

**Comment by riking** [comment]
Can the apiserver block creation of these invalid cross-namespace ownerReferences?

**Comment by lavalamp** [comment]
> Can the apiserver block creation of these invalid cross-namespace ownerReferences?

No, apiserver doesn't enforce references. For better or worse, but I don't see us starting now.

**Comment by gianarb** [comment]
Hello @liggitt !
Bug Triage team here for the 1.18 release. This is a friendly reminder that code freeze is scheduled for 5 March. Is this issue still intended for milestone 1.18, or should it extend to the next one? Thanks in advance!

**Comment by smourapina** [comment]
As we are now in code freeze for 1.18, I am moving this to the next milestone.
/milestone v1.19

**Comment by barkbay** [comment]
We also hit this one. A user was copying some resources from a namespace to another, keeping the owner reference. When the controller manager restarted the garbage collector also deleted some PVCs in the original namespace 😕 

@lavalamp 
> No, apiserver doesn't enforce references. For better or worse, but I don't see us starting now.

Any particular concern about rejecting creation/update of an object when the ownerReference does not exist in the namespace or at the cluster level ?

@liggitt 
> I agree this is important to fix. I am looking into potential mitigations to make behavior in the presence of these ownerRefs more deterministic

Could you share your thoughts ? At first I would say that a workaround would be to not only use the UID in the `absentOwnerCache` but also the namespace _(for namespace scoped resources ofc)_ . It doesn't feel like a "clean" solution but it would prevent this issue ?

**Comment by jtslear** [comment]
Hello, @jtslear here from Bug Triage! This issue has not been updated for a long time, so I'd like to check on the status. The code freeze is starting June 25th and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.

As the issue is tagged for 1.19, is it still planned for this release?

**Comment by lavalamp** [comment]
> Any particular concern about rejecting creation/update of an object when the ownerReference does not exist in the namespace or at the cluster level ?

Yes, it's much more complicated than it sounds like, e.g. the owning object might not exist in the apiserver, it might be in e.g. a different aggregated apiserver, which the apiserver handling the request doesn't know about.

**Comment by jtslear** [comment]
@kubernetes/sig-api-machinery-bugs 

I haven't seen a followup on this issue for awhile.  Is this still planned for the v1.19 release?

**Comment by caesarxuchao** [comment]
https://github.com/kubernetes/kubernetes/pull/78990 tried to add `resource` and `namespace` field to the `ownerReference`.  It brings us one step closer to fix this issue. @deads2k do you have plan to revive that PR?

**Comment by liggitt** [comment]
> Could you share your thoughts ? At first I would say that a workaround would be to not only use the UID in the absentOwnerCache but also the namespace (for namespace scoped resources ofc) . It doesn't feel like a "clean" solution but it would prevent this issue ?

Something like that, probably the namespace+group+kind+uid coordinates. I had a quick proof of concept around this but hadn't satisfied myself that it behaved well in all conditions.

**Comment by caesarxuchao** [comment]
Based on https://github.com/kubernetes/kubernetes/issues/65200#issuecomment-645180760, I see that we are not trying to validate the ownerReference during create/update, but trying to fix cache of gc to make its behavior deterministic, letting gc immediately delete the object if its ownerRef is pointing at wrong namespace/group/kind.

@liggitt can you share the corner cases when you are back?

**Comment by jtslear** [comment]
Due to reaching the code freeze deadline, I'm going to bump this to the next release.

/milestone v1.20

**Comment by liggitt** [comment]
I'm working on a fix for this for the 1.20 timeframe

**Comment by sayanchowdhury** [comment]
👋🏽 Hey @liggitt ! I'm from the Bug Triage team. This issue has not been updated for a while, so I'd like to check on the status. The code freeze is starting November 12th (about 3 weeks from now) and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.

As the issue is tagged for 1.20, is it still planned for this release?

**Comment by liggitt** [comment]
Yes, work on this is in progress for 1.20

**Comment by liggitt** [comment]
Fixed in 1.20 by https://github.com/kubernetes/kubernetes/pull/92743
